dialer.go

```diff
@@ -68,175 +63,172 @@ func NewDialer(opts *DialerOpts) Dialer {
-	return &dialer{
-		windowSize:            opts.WindowSize,
-		maxPadding:            opts.MaxPadding,
-		maxStreamsPerConn:     opts.MaxStreamsPerConn,
-		maxLiveConns:          opts.MaxLiveConns,
-		idleInterval:          opts.IdleInterval,
-		pingInterval:          opts.PingInterval,
-		redialSessionInterval: opts.RedialSessionInterval,
-		pool:            opts.Pool,
-		cipherCode:      opts.Cipher,
-		serverPublicKey: opts.ServerPublicKey,
-		liveSessions:    liveSessions,
-		numLive:         1, // the nullSession
-		emaRTT:          ema.NewDuration(0, 0.5),
+	var lc ClientLifecycleListener
+	if opts.Lifecycle != nil {
+		lc = opts.Lifecycle
+	} else {
+		lc = NoopClientLifecycleListener()
+	}
+	d := &dialer{
+		windowSize:        opts.WindowSize,
+		maxPadding:        opts.MaxPadding,
+		maxStreamsPerConn: opts.MaxStreamsPerConn,
+		pingInterval:      opts.PingInterval,
+		pool:              opts.Pool,
+		cipherCode:        opts.Cipher,
+		serverPublicKey:   opts.ServerPublicKey,
+		liveSessions:      make(chan sessionIntf, opts.LiveConns),
+		emaRTT:            ema.NewDuration(0, 0.5),
+		dial:              opts.Dial,
+		pendingSessions:   make(chan *pendingSession, opts.LiveConns),
+		lifecyle:          lc,
+		name:              opts.Name,
+	}
+	d.lifecyle.OnStart()
+	for i := 0; i < opts.LiveConns-1; i++ {
+		d.pendingSessions <- newPendingSession(d.name)
+	}
+
+	// We create another background session with a shorter dial timeout to ensure liveness in the case of network
+	// disruptions.
+	d.pendingSessions <- &pendingSession{
+		name:         "liveness to " + d.name,
+		dialTimeout:  5 * time.Second,
+		sleepOnError: 1 * time.Second,
 	}
+	go d.maintainTCPConnections()
+	return d
 }
 
 type dialer struct {
-	windowSize            int
-	maxPadding            int
-	maxLiveConns          int
-	maxStreamsPerConn     uint16
-	idleInterval          time.Duration
-	pingInterval          time.Duration
-	redialSessionInterval time.Duration
-	pool                  BufferPool
-	cipherCode            Cipher
-	serverPublicKey       *rsa.PublicKey
-	muNumLivePending      sync.Mutex
-	numLive               int
-	numPending            int
-	liveSessions          chan sessionIntf
-	emaRTT                *ema.EMA
+	windowSize        int
+	maxPadding        int
+	maxStreamsPerConn uint16
+	pingInterval      time.Duration
+	pool              BufferPool
+	cipherCode        Cipher
+	serverPublicKey   *rsa.PublicKey
+	liveSessions      chan sessionIntf
+	pendingSessions   chan *pendingSession
+	emaRTT            *ema.EMA
+	dial              DialFN
+	lifecyle          ClientLifecycleListener
+	name              string
 }
 
-func (d *dialer) Dial(dial DialFN) (net.Conn, error) {
-	return d.DialContext(context.Background(), dial)
+// maintainTCPConnections maintains background TCP connection(s) and associated lampshade session(s)
+func (d *dialer) maintainTCPConnections() {
+	for rs := range d.pendingSessions {
+		go d.trySession(rs)
```

Comment:
Better to use `ops.Go` to maintain any ops context if needed.

Reply:
will do